server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotTaskRunnerTests.java

```diff
@@ -0,0 +1,195 @@
+        }
+
+        public int finishedShardSnapshots() {
+            return finishedShardSnapshots.get();
+        }
+
+        public int finishedShardSnapshotTasks() {
+            return finishedShardSnapshotTasks.get();
+        }
+    }
+
+    private static BlobStoreIndexShardSnapshot.FileInfo dummyFileInfo() {
+        String filename = randomAlphaOfLength(10);
+        StoreFileMetadata metadata = new StoreFileMetadata(filename, 10, "CHECKSUM", Version.CURRENT.luceneVersion.toString());
+        return new BlobStoreIndexShardSnapshot.FileInfo(filename, metadata, null);
+    }
+
+    private SnapshotShardContext dummyContext() {
+        return dummyContext(new SnapshotId(randomAlphaOfLength(10), UUIDs.randomBase64UUID()), randomMillisUpToYear9999());
+    }
+
+    private SnapshotShardContext dummyContext(final SnapshotId snapshotId, final long startTime) {
+        IndexId indexId = new IndexId(randomAlphaOfLength(10), UUIDs.randomBase64UUID());
+        ShardId shardId = new ShardId(indexId.getName(), indexId.getId(), 1);
+        Settings settings = Settings.builder()
+            .put(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            .build();
+        IndexSettings indexSettings = new IndexSettings(
+            IndexMetadata.builder(indexId.getName()).settings(settings).build(),
+            Settings.EMPTY
+        );
+        Store dummyStore = new Store(shardId, indexSettings, new ByteBuffersDirectory(), new DummyShardLock(shardId));
+        return new SnapshotShardContext(
+            dummyStore,
+            null,
+            snapshotId,
+            indexId,
+            new Engine.IndexCommitRef(null, () -> {}),
+            null,
+            IndexShardSnapshotStatus.newInitializing(null),
+            Version.CURRENT,
+            Collections.emptyMap(),
+            startTime,
+            ActionListener.noop()
+        );
+    }
+
+    public void testShardSnapshotTaskRunner() throws Exception {
+        int maxTasks = randomIntBetween(1, threadPool.info(ThreadPool.Names.SNAPSHOT).getMax());
+        MockedRepo repo = new MockedRepo();
+        ShardSnapshotTaskRunner taskRunner = new ShardSnapshotTaskRunner(maxTasks, executor, repo::snapshotShard, repo::snapshotFile);
+        repo.setTaskRunner(taskRunner);
+        int enqueuedSnapshots = randomIntBetween(maxTasks * 2, maxTasks * 10);
+        for (int i = 0; i < enqueuedSnapshots; i++) {
+            threadPool.generic().execute(() -> taskRunner.enqueueShardSnapshot(dummyContext()));
+        }
+        // Eventually all snapshots are finished
+        assertBusy(() -> {
+            assertThat(repo.finishedShardSnapshots(), equalTo(enqueuedSnapshots));
+            assertThat(taskRunner.runningTasks(), equalTo(0));
+        });
+        assertThat(taskRunner.queueSize(), equalTo(0));
```

Comment:
I wonder if there is a risk of the `assertBusy` exiting too early, since `runningTasks` could in theory be 0 with many more queued items. Suppose all your runners has just done the decrement and now wants to do the increment. At that time, the `runningTasks()` could be 0?

You will likely have to include either the queue size or the file-snapshot task check in the assert busy. If you include queue-size it must be checked before the `runningTasks` check.

Reply:
Maybe there is a misunderstanding but `finishedShardSnapshots` means all file/meta tasks are finished, not just the meta work. I think the assertBusy there waits until all tasks (of both type) are finished.